Add short range neighbor methods

jhossbach opened the pull request:
Fixes #4399, fixes #4438

Description of changes:
- `CellStructure.cpp`: Added the `CellStructure::run_on_particle_short_range_neighbors()` method which runs a kernel function over all particles inside the cell and its neighbors of a given particle
- `cells.hpp`: Added the `mpi_get_short_range_neighbors()` function to execute a parallel search
   - can be called from the Python interface via `system.cell_system.get_neighbors.get_neighbors(p, distance)`
- `energy.hpp`: Added the `compute_non_bonded_pair_energy()` function which returns both the short-range Coulomb interactions plus the non-bonded energy contributions  of two particles
   - can be called from the Python interface via `system.analysis.particle_energy(p)`
- Reaction methods can use the new neighbor search algorithm with constructor argument `search_algorithm="parallel"` (default is `search_algorithm="order_n"`); on 1 MPI rank the original order N algorithm is faster since the parallel algorithm introduces some overhead due to the ghost update (this overhead is negligible with 2 or more MPI ranks)

jhossbach commented:
The most recent version can now be found here: https://github.com/jhossbach/espresso/tree/short_range_neighbors_patch
@stekajack: Most tests in the unit test produce the expected results now, but there is still a problem with PBC neighbors which I will look at during the next coding day.

jhossbach commented:
## Left ToDo's :

- [x] Write `unit_tests/short_range_neighbors_test.cpp` to test that neighbor cells are chosen correctly
- [x] Add python interface for the short range neighbors (see #4399)
- [x] Add python interface for the short range energy contribution
- [ ] ~Test short range neighbors via python interface (or as a cpp unit test within in `unit_tests/short_range_neighbors_test.cpp`)~
- [x] Test `particle_short_range_energy_contribution` within python interface

jhossbach commented:
One of the changes introduced an error with the mpi_... functions, I will check and push an updated version.

jngrad commented on src/python/espressomd/cellsystem.pyx:
```diff
@@ -129,6 +130,27 @@ cdef class CellSystem:
+
+        Returns
+        -------
+        list of tuples of :obj:`int`
```
```suggestion
        (N,) array_like of :obj:`int`
```

jngrad commented on src/python/espressomd/cellsystem.pyx:
```diff
@@ -129,6 +130,27 @@ cdef class CellSystem:
+
+        Parameters
+        ----------
+        particle : :attr:`~espressomd.particle_data.ParticleHandle.type`
```
```suggestion
        particle : :class:`~espressomd.particle_data.ParticleHandle`
```

jngrad commented on src/python/espressomd/cellsystem.pyx:
```diff
@@ -129,6 +130,27 @@ cdef class CellSystem:
         handle_errors("Error in get_pairs()")
         return pairs
 
+    def get_neighbors(self, ParticleHandle particle, distance):
```
```suggestion
    def get_neighbors(self, particle, distance):
```
Also don't forget `utils.check_type_or_throw_except()` on both `distance` and `particle` to make sure we pass to correct types to the C++ function.

jngrad commented on src/python/espressomd/cellsystem.pxd:
```diff
@@ -39,6 +39,11 @@ cdef extern from "CellStructureType.hpp":
         CELL_STRUCTURE_REGULAR "CellStructureType::CELL_STRUCTURE_REGULAR"
         CELL_STRUCTURE_NSQUARE "CellStructureType::CELL_STRUCTURE_NSQUARE"
 
+cdef extern from "particle_data.hpp":
+    ctypedef struct particle "Particle"
+    const particle & get_particle_data(int pid) except +
+
+
```
```suggestion
```

Note: each .pyx file includes all .pxd files. This struct and function are already declared in `particle_data.pxd`, so it's not necessary to include them here again.

jngrad commented on src/python/espressomd/analyze.pyx:
```diff
@@ -344,6 +345,23 @@ class Analysis:
+
+        Returns
+        -------
+        :obj: `float` non-bonded energy of that particle
```
```suggestion
        :obj: `float`
            non-bonded energy of that particle
```

jngrad commented on src/python/espressomd/analyze.pyx:
```diff
@@ -344,6 +345,23 @@ class Analysis:
+
+        Parameters
+        ----------
+        particle : :attr:`~espressomd.particle_data.ParticleHandle.type`
```
```suggestion
        particle : :class:`~espressomd.particle_data.ParticleHandle`
```

jngrad commented on src/python/espressomd/analyze.pyx:
```diff
@@ -344,6 +345,23 @@ class Analysis:
         utils.handle_errors("calculate_energy() failed")
         return obs
 
+    def particle_energy(self, ParticleHandle particle):
```
```suggestion
    def particle_energy(self, particle):
```

jngrad commented on src/core/energy.hpp:
```diff
@@ -42,4 +42,16 @@ double calculate_current_potential_energy_of_system();
+ * energy contribution for particle p
+ *
+ * @param pid Particle id
+ * @return Double containing short-range Coulomb and non-bonded energy of the
+ * particle
```
Is Coulomb part of `calc_non_bonded_pair_energy()`?

jhossbach replied:
No. Corrected in the wrap up commit

jngrad commented on src/core/energy.cpp:
```diff
@@ -128,3 +128,33 @@ double observable_compute_energy() {
+  if (p) {
+    auto kernel = [&ret](Particle const &p, Particle const &p1,
+                         Utils::Vector3d const &vec) {
+      auto const dist = sqrt(vec.norm2());
+      auto const &ia_params = *get_ia_param(p.p.type, p1.p.type);
+      // Add energy for current particle pair to result
+      ret += calc_non_bonded_pair_energy(p, p1, ia_params, vec, dist);
```
```suggestion
#ifdef EXCLUSIONS
      if (not do_nonbonded(p, p1))
        return;
#endif
      auto const &ia_params = *get_ia_param(p.type(), p1.type());
      // Add energy for current particle pair to result
      ret += calc_non_bonded_pair_energy(p, p1, ia_params, vec, vec.norm());
```

jngrad commented on src/core/cells.cpp:
```diff
@@ -111,6 +112,38 @@ std::vector<PairInfo> non_bonded_loop_trace() {
+  auto kernel = [&ret, cutoff2](Particle const &p, Particle const &p1,
+                                Utils::Vector3d const &vec) {
+    if (vec.norm2() < cutoff2) {
+      ret.emplace_back(p1.p.identity);
+    };
```
```suggestion
      ret.emplace_back(p1.id());
    }
```

jngrad commented on src/core/CellStructure.hpp:
```diff
@@ -663,6 +663,58 @@ struct CellStructure {
+
+private:
+  template <class Kernel, class DistanceFunc>
+  void short_range_neighbor_loop(Particle const &p, Cell *cell, Kernel kernel,
+                                 DistanceFunc df) {
```
```suggestion
  void short_range_neighbor_loop(Particle const &p, Cell *const cell,
                                 Kernel &kernel, DistanceFunc const &df) {
```

jngrad commented on src/core/CellStructure.hpp:
```diff
@@ -663,6 +663,58 @@ struct CellStructure {
+   * @return false if cell is not found via particle_to_cell, otherwise true
+   */
+  template <class Kernel>
+  bool run_on_particle_short_range_neighbors(Particle const &p, Kernel kernel) {
```
```suggestion
  bool run_on_particle_short_range_neighbors(Particle const &p,
                                             Kernel &kernel) {
```

jngrad commented on src/core/CellStructure.hpp:
```diff
@@ -663,6 +663,58 @@ struct CellStructure {
+   * @brief Run kernel on all particles inside cell and neighbors.
+   *
+   * @param p Particle to identify cell and neighbors
+   * @param kernel Function with (Particle, Particle, Vector)
```
```suggestion
   * @param kernel Function with signature <tt>double(Particle, Particle, Vector)</tt>
```

jngrad commented on src/core/CellStructure.hpp:
```diff
@@ -663,6 +663,58 @@ struct CellStructure {
+   *
+   * @param p Particle to identify cell and neighbors
+   * @param kernel Function with (Particle, Particle, Vector)
+   * @return false if cell is not found via particle_to_cell, otherwise true
```
```suggestion
   * @return false if cell is not found via @ref particle_to_cell, otherwise true
```

jngrad commented on src/core/CellStructure.hpp:
```diff
@@ -663,6 +663,58 @@ struct CellStructure {
+                                 DistanceFunc df) {
+    /* Iterate over particles inside cell */
+    for (auto const &part : cell->particles()) {
+      if (&part == &p) {
```
would this equality check work with `part.id() == p.id()`?

jhossbach replied:
I was following https://github.com/espressomd/espresso/issues/4399#issue-1072323710, but using `id()` looks more elegant to me.
Is there any difference between `.id()` and `.identity()`?

jngrad replied:
Both methods are identical, we'll most likely remove one of them soon.
Actually here you could even write `if (part == p)` thanks to:
https://github.com/espressomd/espresso/blob/a91b0dc6bbbaee101eeb5d93ec36ec378483b350/src/core/Particle.hpp#L408-L410

jngrad commented on src/core/cells.cpp:
```diff
@@ -28,6 +28,7 @@
 
 #include "Particle.hpp"
 #include "communication.hpp"
+#include "energy_inline.hpp"
```
```suggestion
```

jngrad commented on src/core/CellStructure.hpp:
```diff
@@ -663,6 +663,61 @@ struct CellStructure {
+   * @param p Particle to identify cell and neighbors
+   * @param kernel Function with signature <tt>double(Particle, Particle,
+   * Vector)</tt>
+   * @return false if cell is not found via @ref particle_to_cell, otherwise
```
```suggestion
   * @return false if cell is not found via @c particle_to_cell, otherwise
```
sorry, I forgot that private methods couldn't be referenced in Doxygen

jngrad reviewed (changes requested):
Looks good to me! Once the remaining proposed changes are in, we can move forward with the merge.

jhossbach commented:
There is a problem with getting the right distances when using more than one MPI rank, this needs to be adressed first

jngrad commented:
@RudolfWeeber could you please have a look at this PR? Somehow running a short-range loop (e.g. to get the list of neighbors) updates the ghost information, but only in the cellsystem, so we end up with a corrupted particle node cache. This is why running analysis functions that act on particle positions after particles have been moved return results based on the old position, unless we run the integrator with 0 steps.

To reproduce the issue, run the test on 3 MPI ranks. This triggers a segfault due to a corrupted particle node cache.

When replacing
```c++
  if (pnode == this_node) {
    assert(cell_structure.get_local_particle(part));
    return *cell_structure.get_local_particle(part);
  }
```
by
```c++
  if (pnode == this_node) {
    auto const p_ptr = cell_structure.get_local_particle(part);
    if (p_ptr) {
      return *p_ptr;
    }
  }
```
in `src/core/particle_data.cpp:get_particle_data()`, the segfault moves to the particle insertion code. When commenting out the block
```c++
  if (cell_structure.get_resort_particles()) {
    cells_update_ghosts(global_ghost_flags());
  }
```
that was introduced by this PR in `src/core/cells.cpp:mpi_get_short_range_neighbors_local()`, the segfault persists.

The more fundamental issue here is that doing
```python
system.part.all().pos = np.random.random((20, 3)) * system.box_l
```
does **not** update the particle properties, but instead queues the change until the next integration loop. So at the minimum one has to do
```python
system.integrator.run(0, recalc_forces=False)
```
before attempting to update particle positions again. This also explains why the `ParticleHandle` property `node` still returns the old MPI rank after moving the particle to a different MPI domain, until the next integration.

RudolfWeeber commented:
Please try `on_observable_calc()` from event.cpp rather than the manual resort call at the beginning of the neigbor search.

RudolfWeeber commented:
The resort which was already present means that particles can move from one node to an other. Then the particle node cache needs to be invalidated, which is done in the above mentioned funciton.

jhossbach commented:
The assertion in the `check_neighbors` function fails when running with 5 and 7 MPI cores. Also, we should implement a runtime check to ensure that the distance is less than the cell size, otherwise particles that are within distance but not in the neighborlist of that cell will not be included

jhossbach committed:
core: Efficient search algorithm for pair interactions

Implement parallel kernels to find particle pairs within a cutoff
distance of a specific particle and to calculate the short-range
non-bonded energy of a specific particle.

Co-authored-by: Jean-Noël Grad <jgrad@icp.uni-stuttgart.de>
Co-authored-by: Rudolf Weeber <weeber@icp.uni-stuttgart.de>

jhossbach commented:
I added a sanity check to prevent looking for particles for distances further than the cell size.
I also adapted the unit test, can you have another look @jngrad?

jngrad commented:
The `HybridDecomposition` cell system is now partially supported thanks to @pkreissl. This was achieved by iterating over the N-square cells and injecting the local regular decomposition cell as a neighbor. This means that in a system with ions on a regular decomposition cell system and colloids on a N-square cell system, ions can detect any colloid without any distance restriction, however colloids can only detect ions in the regular cell of the same MPI node. I tried adding the regular cell neighbors on the local node to the local N-square cell, but then I get double counting since the regular cell neighbors already contain the N-square cells. To fix that, one would have to create temporary vectors in the hybrid decomposition constructor. However, there is another issue: since the colloid cell has no geometric boundary, a colloid fixed in space can potentially detect all ions in regular cell of MPI rank 0 at time step 0, and all ions in regular cell of MPI rank 1 at time step 1, if the N-square cells are resorted. To fully support `HybridDecomposition`, one probably would have to add all regular cells as neighbors of the N-square cells.

jngrad committed:
core: Parallel search for reaction methods
